// @fhevm/mock-utils/_cjs/fhevm/contracts/FHEVMExecutor.js
"use strict";
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _FHEVMExecutor_fhevmExecutorContract, _FHEVMExecutor_fhevmExecutorContractAddress, _FHEVMExecutor_aclAddress, _FHEVMExecutor_fheGasLimitAddress, _FHEVMExecutor_inputVerifierAddress, _FHEVMExecutor_version;
Object.defineProperty(exports, "__esModule", { value: true });
exports.FHEVMExecutor = void 0;
const ethers_1 = require("ethers");
const address_js_1 = require("../../utils/address.js");
const error_js_1 = require("../../utils/error.js");
const string_js_1 = require("../../utils/string.js");
const abiFHEVMExecutor = [
    "function getVersion() external pure virtual returns (string memory)",
    "function getInputVerifierAddress() public view virtual returns (address)",
    "function getFHEGasLimitAddress() public view virtual returns (address)",
    "function getACLAddress() public view virtual returns (address)",
];
class FHEVMExecutor {
    constructor(runner, fhevmExecutorContractAddress) {
        _FHEVMExecutor_fhevmExecutorContract.set(this, void 0);
        _FHEVMExecutor_fhevmExecutorContractAddress.set(this, void 0);
        _FHEVMExecutor_aclAddress.set(this, void 0);
        _FHEVMExecutor_fheGasLimitAddress.set(this, void 0);
        _FHEVMExecutor_inputVerifierAddress.set(this, void 0);
        _FHEVMExecutor_version.set(this, void 0);
        (0, address_js_1.assertIsAddress)(fhevmExecutorContractAddress, "fhevmExecutorContractAddress");
        __classPrivateFieldSet(this, _FHEVMExecutor_fhevmExecutorContractAddress, fhevmExecutorContractAddress, "f");
        __classPrivateFieldSet(this, _FHEVMExecutor_fhevmExecutorContract, new ethers_1.ethers.Contract(fhevmExecutorContractAddress, abiFHEVMExecutor, runner), "f");
    }
    get address() {
        return __classPrivateFieldGet(this, _FHEVMExecutor_fhevmExecutorContractAddress, "f");
    }
    get runner() {
        (0, error_js_1.assertFhevm)(__classPrivateFieldGet(this, _FHEVMExecutor_fhevmExecutorContract, "f").runner);
        return __classPrivateFieldGet(this, _FHEVMExecutor_fhevmExecutorContract, "f").runner;
    }
    get version() {
        (0, error_js_1.assertFhevm)(__classPrivateFieldGet(this, _FHEVMExecutor_version, "f") !== undefined, `FHEVMExecutor wrapper is not yet initialized`);
        return __classPrivateFieldGet(this, _FHEVMExecutor_version, "f");
    }
    get aclAddress() {
        (0, error_js_1.assertFhevm)(__classPrivateFieldGet(this, _FHEVMExecutor_aclAddress, "f") !== undefined, `FHEVMExecutor wrapper is not yet initialized`);
        return __classPrivateFieldGet(this, _FHEVMExecutor_aclAddress, "f");
    }
    get fheGasLimitAddress() {
        (0, error_js_1.assertFhevm)(__classPrivateFieldGet(this, _FHEVMExecutor_fheGasLimitAddress, "f") !== undefined, `FHEVMExecutor wrapper is not yet initialized`);
        return __classPrivateFieldGet(this, _FHEVMExecutor_fheGasLimitAddress, "f");
    }
    get inputVerifierAddress() {
        (0, error_js_1.assertFhevm)(__classPrivateFieldGet(this, _FHEVMExecutor_inputVerifierAddress, "f") !== undefined, `FHEVMExecutor wrapper is not yet initialized`);
        return __classPrivateFieldGet(this, _FHEVMExecutor_inputVerifierAddress, "f");
    }
    static async create(runner, fhevmExecutorContractAddress) {
        const fhevmExecutor = new FHEVMExecutor(runner, fhevmExecutorContractAddress);
        await fhevmExecutor.initialize();
        return fhevmExecutor;
    }
    async initialize() {
        (0, error_js_1.assertFhevm)(__classPrivateFieldGet(this, _FHEVMExecutor_aclAddress, "f") === undefined, `FHEVMExecutor wrapper already initialized`);
        (0, error_js_1.assertFhevm)(__classPrivateFieldGet(this, _FHEVMExecutor_fheGasLimitAddress, "f") === undefined, `FHEVMExecutor wrapper already initialized`);
        (0, error_js_1.assertFhevm)(__classPrivateFieldGet(this, _FHEVMExecutor_inputVerifierAddress, "f") === undefined, `FHEVMExecutor wrapper already initialized`);
        (0, error_js_1.assertFhevm)(__classPrivateFieldGet(this, _FHEVMExecutor_version, "f") === undefined, `FHEVMExecutor wrapper already initialized`);
        __classPrivateFieldSet(this, _FHEVMExecutor_aclAddress, await __classPrivateFieldGet(this, _FHEVMExecutor_fhevmExecutorContract, "f").getACLAddress(), "f");
        (0, address_js_1.assertIsAddress)(__classPrivateFieldGet(this, _FHEVMExecutor_aclAddress, "f"), "aclAddress");
        __classPrivateFieldSet(this, _FHEVMExecutor_fheGasLimitAddress, await __classPrivateFieldGet(this, _FHEVMExecutor_fhevmExecutorContract, "f").getFHEGasLimitAddress(), "f");
        (0, address_js_1.assertIsAddress)(__classPrivateFieldGet(this, _FHEVMExecutor_fheGasLimitAddress, "f"), "fheGasLimitAddress");
        __classPrivateFieldSet(this, _FHEVMExecutor_inputVerifierAddress, await __classPrivateFieldGet(this, _FHEVMExecutor_fhevmExecutorContract, "f").getInputVerifierAddress(), "f");
        (0, address_js_1.assertIsAddress)(__classPrivateFieldGet(this, _FHEVMExecutor_inputVerifierAddress, "f"), "inputVerifierAddress");
        __classPrivateFieldSet(this, _FHEVMExecutor_version, await __classPrivateFieldGet(this, _FHEVMExecutor_fhevmExecutorContract, "f").getVersion(), "f");
        (0, string_js_1.assertIsString)(__classPrivateFieldGet(this, _FHEVMExecutor_version, "f"), "version");
    }
}
exports.FHEVMExecutor = FHEVMExecutor;
_FHEVMExecutor_fhevmExecutorContract = new WeakMap(), _FHEVMExecutor_fhevmExecutorContractAddress = new WeakMap(), _FHEVMExecutor_aclAddress = new WeakMap(), _FHEVMExecutor_fheGasLimitAddress = new WeakMap(), _FHEVMExecutor_inputVerifierAddress = new WeakMap(), _FHEVMExecutor_version = new WeakMap();
