// @fhevm/mock-utils/_cjs/ethers/provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.minimalProviderSend = minimalProviderSend;
exports.getProviderChainId = getProviderChainId;
exports.getSignerChainId = getSignerChainId;
exports.canSign = canSign;
exports.isReadonlyContract = isReadonlyContract;
exports.isReadonlyProvider = isReadonlyProvider;
const error_js_1 = require("../utils/error.js");
async function minimalProviderSend(provider, method, params) {
    let response;
    if ("request" in provider && typeof provider.request === "function") {
        response = await provider.request({ method, params });
    }
    else if ("send" in provider && typeof provider.send === "function") {
        response = await provider.send(method, params);
    }
    else {
        throw new Error("Invalid provider: must implement request() or send()");
    }
    return response;
}
async function getProviderChainId(provider) {
    const chainIdHex = await minimalProviderSend(provider, "eth_chainId", []);
    return Number(BigInt(chainIdHex));
}
async function getSignerChainId(signer) {
    const provider = signer.provider;
    if (!provider) {
        throw new error_js_1.FhevmError("Unable to determine signer provider");
    }
    const network = await provider.getNetwork();
    return Number(network.chainId);
}
function canSign(obj) {
    if (!obj) {
        throw new error_js_1.FhevmError(`Invalid argument`);
    }
    const isDirectSigner = typeof obj.signTransaction === "function";
    const canProduceSigner = typeof obj.getSigner === "function";
    return isDirectSigner || canProduceSigner;
}
function isReadonlyContract(contract) {
    return !canSign(contract.runner);
}
function isReadonlyProvider(obj) {
    if (!obj) {
        throw new error_js_1.FhevmError(`Invalid argument`);
    }
    return (!canSign(obj) &&
        typeof obj.estimateGas === "function" &&
        typeof obj.call === "function" &&
        typeof obj.getBlock === "function" &&
        typeof obj.getNetwork === "function" &&
        typeof obj.getCode === "function");
}
